usr/lib/common/buffer.c

```diff
@@ -0,0 +1,108 @@
+/*
+ * COPYRIGHT (c) International Business Machines Corp. 2022
+ *
+ * This program is provided under the terms of the Common Public License,
+ * version 1.0 (CPL-1.0). Any use, reproduction or distribution for this
+ * software constitutes recipient's acceptance of CPL-1.0 terms which can be
+ * found in the file LICENSE file or at
+ * https://opensource.org/licenses/cpl1.0.php
+ */
+
+#include <stdarg.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include <buffer.h>
+
+#define BUFFER_INIT_SIZE	(128ul)
+
+struct _buffer {
+    size_t b_size;
+    char *b;
+};
+
+p11_buffer_t *p11_buffer_new(void)
+{
+    p11_buffer_t *ret = malloc(sizeof(p11_buffer_t));
+    if (ret == NULL)
+        goto out;
+
+    ret->b_size = BUFFER_INIT_SIZE;
+    ret->b = calloc(BUFFER_INIT_SIZE, sizeof(char));
+
+    if (ret->b == NULL)
+        goto out_free;
+
+    return ret;
+
+out_free:
+    free(ret);
+out:
+    return NULL;
+}
+
+void p11_buffer_free(p11_buffer_t *buf)
+{
+    free(buf->b);
+    free(buf);
+}
+
+void p11_buffer_reset(p11_buffer_t *buf)
+{
+    memset(buf->b, 0, buf->b_size);
+}
+
+const char *p11_buffer_char(const p11_buffer_t *buf)
+{
+    return (const char *) buf->b;
+}
+
+size_t p11_buffer_size(const p11_buffer_t *buf)
+{
+    return buf->b_size;
+}
+
+int p11_buffer_append_len(p11_buffer_t *buf, const char *s, size_t len)
+{
+    size_t new_len = strlen(buf->b) + len;
+
+    if (!s)
+        return strlen(buf->b);
+
+    /* extend buffer if required */
+    if ((new_len) > (buf->b_size - 1)) {
+        size_t new_b_size =
+            (((new_len) / BUFFER_INIT_SIZE) + 1) *BUFFER_INIT_SIZE;
+        char *new_b = realloc(buf->b, new_b_size);
+
+        /* ENOMEM */
+        if (new_b == NULL)
+            return 0;
+
+        buf->b = new_b;
+        buf->b_size = new_b_size;
+    }
+
+    strncat(buf->b, s, len);
+    return new_len;
+}
+
+int p11_buffer_append(p11_buffer_t *buf, const char *s)
+{
+    return p11_buffer_append_len(buf, s, strlen(s));
+}
+
+int p11_buffer_printf(p11_buffer_t *buf, const char *fmt, ...)
```

Comment:
if I understand the function correctly, this appends the result of the printf result to the buffer. Should the function then better be named something including 'append', i.e. p11_buffer_append_printf ? 
From the name p11_buffer_printf I originally thought, it would print the buffer contents.....

Reply:
It is a printf() to a buffer. As you proposed, I will rename it to p11_buffer_append_printf().